x/rewards/keeper/providers.go

```diff
@@ -39,40 +51,68 @@ func (k Keeper) distributeMonthlyBonusRewards(ctx sdk.Context) {
 		for _, basepay := range basepays {
-			// calculate the providers bonus base on adjusted base pay
-			reward := specTotalPayout.Mul(basepay.TotalAdjusted).QuoInt(totalbasepay).TruncateInt()
-			totalRewarded = totalRewarded.Add(reward)
-			if totalRewarded.GT(total) {
-				utils.LavaFormatError("provider rewards are larger than the distribution pool balance", nil,
-					utils.LogAttr("distribution_pool_balance", total.String()),
-					utils.LogAttr("provider_reward", totalRewarded.String()))
-				details["error"] = "provider rewards are larger than the distribution pool balance"
-				return
-			}
-			// now give the reward the provider contributor and delegators
-			providerAddr, err := sdk.AccAddressFromBech32(basepay.Provider)
-			if err != nil {
-				continue
-			}
-			_, _, err = k.dualstakingKeeper.RewardProvidersAndDelegators(ctx, providerAddr, basepay.ChainID, reward, string(types.ProviderRewardsDistributionPool), false, false, false)
-			if err != nil {
-				utils.LavaFormatError("failed to send bonus rewards to provider", err, utils.LogAttr("provider", basepay.Provider))
+			if !specTotalPayout.IsZero() {
+				// calculate the providers bonus base on adjusted base pay
+				reward := specTotalPayout.Mul(basepay.TotalAdjusted).QuoInt(totalbasepay).TruncateInt()
+				totalRewarded = totalRewarded.Add(reward)
+				if totalRewarded.GT(total) {
+					utils.LavaFormatError("provider rewards are larger than the distribution pool balance", nil,
+						utils.LogAttr("distribution_pool_balance", total.String()),
+						utils.LogAttr("provider_reward", totalRewarded.String()))
+					details["error"] = "provider rewards are larger than the distribution pool balance"
+					return
+				}
+				// now give the reward the provider contributor and delegators
+				providerAddr, err := sdk.AccAddressFromBech32(basepay.Provider)
+				if err != nil {
+					continue
+				}
+				_, _, err = k.dualstakingKeeper.RewardProvidersAndDelegators(ctx, providerAddr, basepay.ChainID, sdk.NewCoins(sdk.NewCoin(k.stakingKeeper.BondDenom(ctx), reward)), string(types.ProviderRewardsDistributionPool), false, false, false)
+				if err != nil {
+					utils.LavaFormatError("failed to send bonus rewards to provider", err, utils.LogAttr("provider", basepay.Provider))
+				}
+
+				details[providerAddr.String()+" "+spec.ChainID] = reward.String()
 			}
 
-			details[providerAddr.String()+" "+spec.ChainID] = reward.String()
+			// count iprpc cu
+			k.countIprpcCu(specCuMap, basepay.IprpcCu, spec.ChainID, basepay.Provider)
 		}
 	}
+
+	// Get current month IprpcReward and use it to distribute rewards
+	iprpcReward, found := k.PopIprpcReward(ctx, true)
+	if !found {
+		utils.LavaFormatError("current month iprpc reward not found", fmt.Errorf("did not reward providers IPRPC bonus"))
+		return
+	}
+	k.RemoveIprpcReward(ctx, iprpcReward.Id)
+
+	// none of the providers will get the IPRPC reward this month, transfer the funds to the next month
+	if len(specCuMap) == 0 {
```

Comment:
move this to k.distributeIprpcRewards(ctx, iprpcReward, specCuMap)
than everything related to IPRPC will be in one place

Reply:
done. see https://github.com/lavanet/lava/pull/1223/commits/5cca4e2ee138e807c14c67d30d113e3310eb4898